3.2.0 - Typechecking rejecting `toEqual` on Map with values mixing expect.any and real values

### Describe the bug

While this code used to be legit in 3.1.x, it started not to be in 3.2.0 (possibly due to https://github.com/vitest-dev/vitest/pull/7016).

```ts
// Does not compile
expect(new Map<string, number>()).toEqual(
  new Map([
    ['a', 1],
    ['b', expect.any(Number)],
  ])
);

// Compiles
expect(new Map<string, number>()).toEqual(new Map([['a', 1]]));

// Compiles
expect(new Map<string, number>()).toEqual(new Map([['b', expect.any(Number)]]));
```

As it does not seem to be related to #8081 or #8079 I opened this dedicated issue.

### Reproduction

https://stackblitz.com/edit/vitest-dev-vitest-q2nrapfj?file=test%2Fbasic.test.ts

### System Info

```shell
System:
    OS: Windows 11 10.0.26100
    CPU: (12) arm64 Snapdragon(R) X Elite - X1E80100 - Qualcomm(R) Oryon(TM) CPU
    Memory: 37.08 GB / 63.49 GB
  Binaries:
    Node: 22.14.0 - C:\Program Files\nodejs\node.EXE
    npm: 10.9.2 - C:\Program Files\nodejs\npm.CMD
    pnpm: 10.6.5 - ~\AppData\Roaming\npm\pnpm.CMD
  Browsers:
    Edge: Chromium (133.0.3065.69)
  npmPackages:
    @vitest/coverage-v8: ^3.1.4 => 3.1.4
    vitest: ^3.2.0 => 3.2.0
```

### Used Package Manager

npm

### Validations

- [x] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [x] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [x] Read the [docs](https://vitest.dev/guide/).
- [x] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [x] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [x] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

I'm getting similar issue when calling `expect.objectContaining`:

`error TS2345: Argument of type 'Partial<T>' is not assignable to parameter of type 'DeeplyAllowMatchers<T>'`

Because `expect.objectContaining` is a pain to type and takes up a lot of space. I have a helper:

```ts
const oc = <T>(obj: T) => expect.objectContaining(obj);
const ac = <T>(a: Array<T>) => expect.arrayContaining(a);
```

This no longer compiles.

--- a/packages/expect/src/types.ts
+++ b/packages/expect/src/types.ts
@@ -108,8 +108,8 @@ export interface ExpectStatic
   AsymmetricMatchersContaining {
   <T>(actual: T, message?: string): Assertion<T>
   extend: (expects: MatchersObject) => void
-  anything: () => AsymmetricMatcher<unknown>
-  any: (constructor: unknown) => AsymmetricMatcher<unknown>
+  anything: () => any
+  any: (constructor: unknown) => any
   getState: () => MatcherState
   setState: (state: Partial<MatcherState>) => void
   not: AsymmetricMatchersContaining
@@ -202,7 +202,7 @@ export interface JestAssertion<T = any> extends jest.Matchers<void, T>, CustomMa
    * @example
    * expect(user).toEqual({ name: 'Alice', age: 30 });
    */
-  toEqual: <E>(expected: DeeplyAllowMatchers<E>) => void
+  toEqual: <E>(expected: E) => void
 
   /**
    * Use to test that objects have the same types as well as structure.
@@ -210,7 +210,7 @@ export interface JestAssertion<T = any> extends jest.Matchers<void, T>, CustomMa
    * @example
    * expect(user).toStrictEqual({ name: 'Alice', age: 30 });
    */
-  toStrictEqual: <E>(expected: DeeplyAllowMatchers<E>) => void
+  toStrictEqual: <E>(expected: E) => void
 
   /**
    * Checks that a value is what you expect. It calls `Object.is` to compare values.
@@ -240,7 +240,7 @@ export interface JestAssertion<T = any> extends jest.Matchers<void, T>, CustomMa
    *   address: { city: 'Wonderland' }
    * });
    */
-  toMatchObject: <E extends object | any[]>(expected: DeeplyAllowMatchers<E>) => void
+  toMatchObject: <E extends object | any[]>(expected: E) => void
 
   /**
    * Used when you want to check that an item is in a list.

